cmd/govim: check quickfix list origin before update

tbruyelle opened the pull request:
QuickfixDiagnostics routine updates the quickfix list each time it
receives new diagnostics from gopls. This can be problematic if the
quickfix list has been populated by an other plugin, because the user
could be using this plugin.

The fix uses the quickfix title to determine if the quickfix is "owned"
by govim or by an other plugin. If the quicfix title is empty (which
means the quickfix has never been used before) or equal to "govim", then
govim considers it can change it, else it considers it cannot.

This condition is ignored when :GOVIMQuickfixDiagnostics is invoked by
the user, making possible for the user to display again the diagnostics
when he wants to.

tbruyelle commented on cmd/govim/quickfix.go:
```diff
@@ -44,6 +44,12 @@ func (v *vimstate) updateQuickfixWithDiagnostics(force bool, wasPrevNotDiagnosti
 	}
+	var qflist qflistProps
+	v.Parse(v.ChannelExpr(`getqflist({"title":1})`), &qflist)
+	if !force && qflist.Title != "" && qflist.Title != "govim" {
```
On thing I would like to improve is to merge this condition and the one from `v.quickfixIsDiagnostics`.

The later is set to false when Govim populates the quickfix with references, but we could use quickfix title also in that case.

I think if I turn `v.quickfixIsDiagnostics` into a function that returns true if the quickfix title is empty or equal to "govim" (maybe rename that to "govim-diagnostics", because we could add an other title for Govim references), then it would clean a little bit the code :

- only one condition here instead of 2
- no more manual `v.quickfixIsDiagnostics = true|false` since it relies on quickfix title

myitcv replied:
Yes I'd be happy to revisit this. Either in this PR or a later one. As you observed, it's rather clumsy now. Vim is essentially the source of truth for what is in the quickfix window at any given time, and your change helps move us closer to that arrangement.

myitcv commented on cmd/govim/testdata/scenario_default/quickfix_from_other.txt:
```diff
@@ -0,0 +1,95 @@
+# Test that the quickfix window isn't updated by govim if it isn't populated by govim
+
+vim ex 'e main.go'
+vim ex 'call setqflist([{\"filename\":\"foo\"}, {\"filename\":\"bar\"}, {\"filename\":\"baz\"}], \"r\")'
```
This is a great test.

One thing I think we need to do before calling this is verify that the quickfix list gets populated with the error diagnostics from `gopls`. That's not essential to this test, but instead a more realistic reflection of real life. So that way you end up with a call to `vimexprwait` before this call, then as you have it another one afterwards.

tbruyelle replied:
@myitcv OK, I added an assertion for the first error 35c475d

myitcv commented on cmd/govim/testdata/scenario_default/quickfix_from_other.txt:
```diff
@@ -0,0 +1,95 @@
+vim ex 'e main.go'
+vim ex 'call setqflist([{\"filename\":\"foo\"}, {\"filename\":\"bar\"}, {\"filename\":\"baz\"}], \"r\")'
+vimexprwait errors.other GOVIMTest_getqflist()
+# Duplicate error, the quickfix should remain because it hasn't been populated by govim
```
If you put a blank line before this comment, [`testscript`](https://pkg.go.dev/github.com/rogpeppe/go-internal/testscript) will treat it as a section. In the verbose output from `testscript`, sections that complete "successfully" are collapse under the "heading" of the comment that precedes the contiguous lines of that section.

tbruyelle replied:
done 35c475d

myitcv commented on cmd/govim/testdata/scenario_default/quickfix_from_other.txt:
```diff
@@ -0,0 +1,95 @@
+vim ex 'call cursor(6, 1)'
+vim ex 'call feedkeys(\"yyp\")'
+vimexprwait errors.other GOVIMTest_getqflist()
+# Force display diagnostics now
```
Ditto.

tbruyelle replied:
done 35c475d

myitcv reviewed (commented):
Generally LGTM with a couple of suggestions

cc @leitzler for thoughts too.

myitcv reviewed (approved):
LGTM - will leave it up to you @tbruyelle if you want to make the change you discussed in this PR or a follow up.

Thank you for your patience with this change!

tbruyelle commented:
@myitcv thx for the review. I'll try to see if I can implement the change we discuss simply, if yes I'll add a commit, if not I'll probably go with an other MR.

myitcv commented on cmd/govim/quickfix.go:
```diff
@@ -24,32 +28,29 @@ func (q quickfixEntry) equalModuloBuffer(q2 quickfixEntry) bool {
-	v.lastDiagnosticsQuickfix = diagsRef
-	if (!force && !work) || !v.quickfixIsDiagnostics {
+
+	var (
```
Please change this from a `var` block to short variable declarations.

tbruyelle replied:
@myitcv sure no problem, by curiosity why do you prefer this kind of declaration ? On my side I like to pack variable declarations like that, because that formats the code and I found it's more clear and readable.

myitcv replied:
I've honestly not seen much (any?) use of parenthesised `var` declaration lists "outside" of top level file declarations. At the file level they do have things to stand out more, and when combined with `import` and `const` declarations, visually make sense. But within a function, they're a bit clumsy and unnecessarily verbose to my mind: which is precisely what short variable declarations exist to help with (amongst other situations like `for` loops).

tbruyelle replied:
Ok, as you wish b65554c

myitcv reviewed (commented):
Thanks @tbruyelle - broadly speaking this looks fine, with the one requested changed for short variable declarations please. 

However, I'd like @leitzler to opine too, because this potentially influences our strategy with regards to a Vim grep workaround.

leitzler commented on cmd/govim/quickfix.go:
```diff
@@ -24,32 +28,29 @@ func (q quickfixEntry) equalModuloBuffer(q2 quickfixEntry) bool {
+		quickfixIsDiagnostics = v.quickfixIsDiagnostics()
+	)
+	v.lastDiagnosticsQuickfix = diags
+	if (!force && !diagsHasChanged) || (!force && !quickfixIsDiagnostics) {
```
Really nits, but I personally think:
```go
if !force && (!diagsHasChanged || !quickfixIsDiagnostics) {
```
would be slightly more readable and consistent with other force checks in the code.

tbruyelle replied:
@leitzler ofc ! I need to revise my boolean algebra courses :)

d3c8ef1

leitzler reviewed (approved):
First of all, nice feature!

When it comes to vimgrep this slightly alters the behaviour (apart from the obvious, that the existing matches doesn't disappear when diagnostics arrive). The behaviour that do change is if the quickfix list is "owned" by someone but empty. 

e.g.
1) There are no govim diagnostics
2) User calls vimgrep but finds no matches, it returns with a `E480: No match: foobar` message
3) The user introduce a new error, but the quickfix remains empty since the title changed, even if it wasn't populated

In my workflow I use `:cw` to open the quickfix list, and it might be confusing if it "doesn't work" as it isn't clear that the title changed at step 2. I think it make sense if `govim` actually did hi-jack the quickfix if the list is empty.

Apart from that and a tiny nit it LGTM.

tbruyelle commented:
> In my workflow I use `:cw` to open the quickfix list, and it might be confusing if it "doesn't work" as it isn't clear that the title changed at step 2. I think it make sense if `govim` actually did hi-jack the quickfix if the list is empty.

@leitzler What if the new `isQuickfixDiagnostics()` function returns true if the quickfix list is empty ?

Maybe we should rename that function seems it's not really `isQuickfixDiagnostics` but more likely `canQuickfixDiagnostics`... WDYT?

tbruyelle commented:
@leitzler @myitcv While I was trying some other simplifications in `updateQuickfixWithDiagnostics`,  I've spotted a bug in this change :

If I remove those lines : 
```diff
 func (v *vimstate) updateQuickfixWithDiagnostics(force bool) error {
-       if !force && (v.config.QuickfixAutoDiagnostics == nil || !*v.config.QuickfixAutoDiagnostics) {
-               return nil
-       }
```

The test ` TestScripts/scenario_default/quickfix_config` is still green, while it should be red, because `quickfixAutoDiagnostics` is ignored.

To make it red as it should be, I have to update the code in `vimstate.go`, which handles the config change of this particular field. Indeed, the code was updating the quickfix list with empty diagnostics when `autoDiagnostics` was set to false, but without setting the quickfix title. If it set the title as expected, the test goes red.

So the fix introduces a new function `setQuickfixDiagnostics` that populates the quickfix with diagnostics passed in parameters and sets the expected title.

See c977523

myitcv commented:
@tbruyelle thanks for making the various changes. I'll hand over to @leitzler at this point for you both to round out the discussion above.

leitzler commented:
> > In my workflow I use `:cw` to open the quickfix list, and it might be confusing if it "doesn't work" as it isn't clear that the title changed at step 2. I think it make sense if `govim` actually did hi-jack the quickfix if the list is empty.
> 
> @leitzler What if the new `quickfixIsDiagnostics()` function returns true if the quickfix list is empty ?

Yes, sounds good.

> Maybe we should rename that function seems it's not really `quickfixIsDiagnostics` but more likely `quickfixCanDiagnostics`... WDYT?

I agree, that would describe the function better.

tbruyelle commented:
> > > In my workflow I use `:cw` to open the quickfix list, and it might be confusing if it "doesn't work" as it isn't clear that the title changed at step 2. I think it make sense if `govim` actually did hi-jack the quickfix if the list is empty.
> > 
> > 
> > @leitzler What if the new `quickfixIsDiagnostics()` function returns true if the quickfix list is empty ?
> 
> Yes, sounds good.
> 
> > Maybe we should rename that function seems it's not really `quickfixIsDiagnostics` but more likely `quickfixCanDiagnostics`... WDYT?
> 
> I agree, that would describe the function better.

@leitzler I have updated the function accordingly, and added a test. See 4edf3be

Now if the quickfix is empty, it can be updated by diagnostics, whatever its title.

leitzler commented:
Looks good to me!

Please squash the commits as well before merge (as per https://github.com/govim/govim/wiki/Contributors#pull-request-contributions) .

Thank you very much for all the updates (and the patience)!

tbruyelle commented:
> Looks good to me!
> 
> Please squash the commits as well before merge (as per https://github.com/govim/govim/wiki/Contributors#pull-request-contributions) .

done 2ed54c7 

> Thank you very much for all the updates (and the patience)!

No problem, it was a pleasure to contribute to this great project :)
